kpi-tracker/src/main.rs

```diff
@@ -79,117 +84,396 @@ struct BlockDetails {
-    /// Table containing relations between contract instances and transactions.
-    transaction_contract_relations: TransactionsContractsTable,
+/// Holds `tokio_postgres::Client` to query the database and
+/// `PreparedStatements` which can be executed with the client.
+struct DBConn {
+    client:   DBClient,
+    prepared: PreparedStatements,
+}
+
+impl DBConn {
+    /// Create new `DBConn` from `tokio_postgres::config::Config`. If
+    /// `try_create_tables` is true, database tables are created using
+    /// `/resources/schema.sql`.
+    async fn create(conn_string: DBConfig, try_create_tables: bool) -> anyhow::Result<Self> {
+        let (client, connection) = conn_string
+            .connect(NoTls)
+            .await
+            .context("Could not create database connection")?;
+
+        tokio::spawn(async move {
+            if let Err(e) = connection.await {
+                eprintln!("Connection error: {}", e); // TODO: log as error.
+            }
+        });
+
+        if try_create_tables {
+            let create_statements = include_str!("../resources/schema.sql");
+            client
+                .batch_execute(create_statements)
+                .await
+                .context("Failed to execute create statements")?;
+        }
+
+        let prepared = PreparedStatements::new(&client).await?;
+        let db_conn = DBConn { client, prepared };
+
+        Ok(db_conn)
+    }
 }
 
 /// Events from individual transactions to store in the database.
 enum BlockEvent {
-    AccountCreation(AccountAddress, AccountDetails),
-    AccountTransaction(
-        TransactionHash,
-        TransactionDetails,
-        Vec<TransactionAccountRelation>,
-        Vec<TransactionContractRelation>,
-    ),
-    ContractModuleDeployment(ModuleRef, ContractModuleDetails),
+    AccountCreation(CanonicalAccountAddress, AccountDetails),
+    AccountTransaction(TransactionHash, TransactionDetails),
+    ContractModuleDeployment(ModuleRef),
     ContractInstantiation(ContractAddress, ContractInstanceDetails),
 }
 
 /// Queries node for account info for the `account` given at the block
 /// represented by the `block_hash`
-fn account_details(
-    block_hash: BlockHash,
-    account_creation_details: &AccountCreationDetails,
-) -> AccountDetails {
+fn account_details(account_creation_details: &AccountCreationDetails) -> AccountDetails {
     let is_initial = match account_creation_details.credential_type {
         CredentialType::Initial { .. } => true,
         CredentialType::Normal { .. } => false,
     };
 
-    AccountDetails {
-        is_initial,
-        block_hash,
-    }
+    AccountDetails { is_initial }
 }
 
 /// Returns accounts on chain at the give `block_hash`
 async fn accounts_in_block(
     node: &mut Client,
     block_hash: BlockHash,
-) -> anyhow::Result<BTreeMap<CanonicalAccountAddress, AccountDetails>> {
+) -> anyhow::Result<HashMap<CanonicalAccountAddress, AccountDetails>> {
```

Comment:
I am not a fan of using hashmaps for this. It means that iteration order will be implementation defined.

In fact I would recommend returning a Vec<(CanonicalAccountAddress, AccountDetails)>  so as not to reorder things needlessly.